src/common/ad_mem.c

```diff
@@ -3,5 +3,1917 @@
+ad_blob_hdl2store(struct ad_blob_handle bh)
+{
+	D_ASSERT(bh.bh_blob);
+	return &bh.bh_blob->bb_store;
+}
+
+void *
+blob_addr2ptr(struct ad_blob *blob, daos_off_t addr)
+{
+	struct ad_page	*pg;
+	daos_off_t	 off;
+
+	off  = addr & ARENA_SIZE_MASK;
+	addr = addr - blob_addr(blob);
+	pg   = &blob->bb_pages[addr >> ARENA_SIZE_BITS];
+	return &pg->pa_rpg[off];
+}
+
+void *
+ad_addr2ptr(struct ad_blob_handle bh, daos_off_t addr)
+{
+	return blob_addr2ptr(bh.bh_blob, addr);
+}
+
+daos_off_t
+blob_ptr2addr(struct ad_blob *blob, void *ptr)
+{
+	struct ad_arena_df *ad;
+	daos_off_t	    off;
+
+	off = (unsigned long)ptr & ARENA_SIZE_MASK;
+	ad = (struct ad_arena_df *)((unsigned long)ptr & ~ARENA_SIZE_MASK);
+	return ad->ad_addr + off;
+}
+
+daos_off_t
+ad_ptr2addr(struct ad_blob_handle bh, void *ptr)
+{
+	return blob_ptr2addr(bh.bh_blob, ptr);
+}
+
+static int
+group_size_cmp(const void *p1, const void *p2)
+{
+	const struct ad_group_df *gd1 = p1;
+	const struct ad_group_df *gd2 = p2;
+	const struct ad_group	 *gp1 = group_df2ptr(gd1);
+	const struct ad_group	 *gp2 = group_df2ptr(gd2);
+	int			  f1;
+	int			  f2;
+
+	if (gd1->gd_unit < gd2->gd_unit)
+		return -1;
+	if (gd1->gd_unit > gd2->gd_unit)
+		return 1;
+
+	f1 = gd1->gd_unit_free - (gp1 ? gp1->gp_unit_rsv : 0);
+	f2 = gd2->gd_unit_free - (gp2 ? gp2->gp_unit_rsv : 0);
+	if (f1 < f2)
+		return -1;
+	if (f1 > f2)
+		return 1;
+
+	/* use address to identify a specified group */
+	if (gd1->gd_addr < gd2->gd_addr)
+		return -1;
+	if (gd1->gd_addr > gd2->gd_addr)
+		return 1;
+	return 0;
+}
+
+static int
+group_addr_cmp(const void *p1, const void *p2)
+{
+	const struct ad_group_df *gd1 = p1;
+	const struct ad_group_df *gd2 = p2;
+
+	if (gd1->gd_addr < gd2->gd_addr)
+		return -1;
+	if (gd1->gd_addr > gd2->gd_addr)
+		return 1;
+
+	/* two groups have the same address? */
+	D_ASSERT(0);
+	return 0;
+}
+
+static int
+arena_locate(struct ad_blob *blob, bool create, uint32_t arena_id, struct ad_arena_df **ad_p)
+{
+	struct ad_blob_df  *bd = blob->bb_df;
+	bool		    created;
+	bool		    reserved;
+
+	if ((arena_id << ARENA_SIZE_BITS) > blob_size(blob))
+		return -DER_INVAL;
+
+	created = isset((uint8_t *)bd->bd_bmap, arena_id);
+	reserved = isset((uint8_t *)blob->bb_bmap_rsv, arena_id);
+	if (create == (created || reserved))
+		return create ? -DER_EXIST : -DER_NONEXIST;
+
+	/* Arena is the header of each page */
+	*ad_p = (struct ad_arena_df *)blob->bb_pages[arena_id].pa_rpg;
+	return 0;
+}
+
+int
+arena_load(struct ad_blob *blob, uint32_t arena_id, struct ad_arena **arena_p)
+{
+	struct ad_arena	    *arena = NULL;
+	struct ad_arena_df  *ad = NULL;
+	struct ad_group_df  *gd;
+	int		     i;
+	int		     rc;
+
+	rc = arena_locate(blob, false, arena_id, &ad);
+	if (rc) {
+		D_ERROR("No available arena\n");
+		return rc;
+	}
+
+	if (ad->ad_magic != ARENA_MAGIC) {
+		D_ERROR("Invalid arena magic: %x/%x\n", ad->ad_magic, ARENA_MAGIC);
+		return -DER_PROTO;
+	}
+
+	if (ad->ad_incarnation != blob_incarnation(blob)) {
```

Comment:
instead of maintaining another index in DRAM, I decided to use blob-map to index almost everything. For example, for reserve() implementation,  I need to maintain an ephemeral bitmap in DRAM  (this bitmap disappear if the engine crashes), as addition of the bitmap in blob. To avoid another index, the ad_group_df has a backpointer to ad_group, the backpointer can be different each time we start the engine (it's never written to WAL or checkpointed intentionally, but I can be written to SSD by other changes made to ad_group_df), so I'm using the incarnation to check if the backpointer is still valid (it's a. up-to-date pointer, or a stale pointer from the previous run of the engine).